File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.cpp
class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        // Sort events by their end times
        sort(events.begin(), events.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[1] < b[1]; // Sort by end time
        });

        int n = events.size();
        vector<int> maxValues(n); // Prefix maximum array
        maxValues[0] = events[0][2]; // Initialize with the first event's value

        // Fill the prefix maximum array
        for (int i = 1; i < n; ++i) {
            maxValues[i] = max(maxValues[i - 1], events[i][2]);
        }

        int maxSum = 0;

        // Iterate through all events
        for (int i = 0; i < n; ++i) {
            // Find the latest non-overlapping event using binary search
            int left = 0, right = i - 1, prev = -1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (events[mid][1] < events[i][0]) { // Check if non-overlapping
                    prev = mid;
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }

            // Calculate maximum sum
            int currSum = events[i][2]; // Take the current event's value
            if (prev != -1) {
                currSum += maxValues[prev]; // Add the best non-overlapping event
            }
            maxSum = max(maxSum, currSum); // Update the maximum sum
        }

        return maxSum;
    }
};

File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.java
class Solution {
    public int maxTwoEvents(int[][] events) {
        int n=events.length;
        Arrays.sort(events, (a,b)-> Integer.compare(a[0],b[0]));
        int[] suffix=new int[n];
        suffix[n-1]=events[n-1][2];
        for(int i=n-2;i>=0;i--){
            suffix[i]=Math.max(suffix[i+1], events[i][2]);
        }

        int maxRes=0;
        for(int i=0;i<n;i++){
            int y=events[i][1], val=events[i][2];
            maxRes=Math.max(maxRes, val);
            int left=i+1, right=n-1;
            while(left<=right){
                int mid=left+(right-left)/2;
                if(events[mid][0]<=y){
                    left=mid+1;
                }else{
                    right=mid-1;
                }
            }
            // for(int j=left;j<n;j++){
            //     maxRes=Math.max(maxRes, val+events[j][2]); // ye basically O(n^2) hi kar raha hai
            // }
            if(left<n){
                maxRes=Math.max(maxRes, val+suffix[left]);
            }
        }
        return maxRes;
    }
}